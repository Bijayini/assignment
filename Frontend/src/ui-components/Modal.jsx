import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import Button from './Button';

class Modal extends Component {
  static propTypes = {
    toggleOverlay: PropTypes.func.isRequired,
    children: PropTypes.node.isRequired,
    buttons: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.string.isRequired,
        type: PropTypes.oneOf(['primary', 'secondary']).isRequired,
        callback: PropTypes.func.isRequired,
      }),
    ).isRequired,
  };

  componentDidMount() {
    document.addEventListener('keydown', this.escButtonHandling);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.escButtonHandling);
  }

  escButtonHandling = event => {
    const { toggleOverlay } = this.props;
    if (event.key === 'Escape' || event.keyCode === 27) {
      toggleOverlay();
    }
  };

  handelOutsideClick = e => {
    const { toggleOverlay } = this.props;

    if (this.node.contains(e.target)) {
      return false;
    }
    toggleOverlay();
  };

  render() {
    const {toggleOverlay, children, buttons } = this.props;

    return ReactDOM.createPortal(
      <div
        className="pop-over-wrapper"
        onClick={this.handelOutsideClick}
        role="alertdialog"
      >
        <div
          className="pop-over"
          ref={node => {
            this.node = node;
          }}
        >
          <a href="#" className="close-icon" onClick={toggleOverlay}>
            &#215;
          </a>
          {children}
          <div className="button-wrapper">
            {buttons &&
            buttons.length &&
            buttons.map((button, index) => (
              <Button
                type="button"
                onClick={button.callback}
                key={button.type}
                variant={button.type}
                size={buttons.length === 2 ? 'medium' : 'large'}
                classList={index === 0 ? 'left-button' : ''}
              >
                {button.label}
              </Button>
            ))}
          </div>
        </div>
      </div>,
      document.body
    );
  }
}

export default Modal;
