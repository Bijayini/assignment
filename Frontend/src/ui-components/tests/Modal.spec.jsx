import React from 'react';
import { shallow,mount } from 'enzyme';

import Modal from '../Modal';

describe('Modal', () => {
  const toggleOverlayMock = jest.fn();

  const props = {
    toggleOverlay: toggleOverlayMock,
    children: <div className="some-name">Some text</div>,
    buttons: [
      {
        label: 'Some Button',
        type: 'primary',
        callback: jest.fn(),
      },
      {
        label: 'cancel',
        type: 'secondary',
        callback: jest.fn(),
      },
    ],
  };

  it('should match with snapshot', () => {
    const component = shallow(<Modal {...props} />);

    expect(component.debug()).toMatchSnapshot();
  });

  it('should not have button wrapper', () => {
    const newProps = {
      ...props,
      buttons: [],
    };
    const component = shallow(<Modal {...newProps} />);

    expect(component.find('ButtonWithTheme').length).toBe(0);
  });

  describe('handelOutsideClick', () => {
    const component = mount(<Modal {...props} />);

    const modal = component.find('Modal');

    it('should not close modal on clicking on modal box', () => {
      modal.find('.pop-over').simulate('click');

      expect(toggleOverlayMock).not.toHaveBeenCalled();
    });

    it('should close modal on clicking on modal box wrapper', () => {
      modal.find('.pop-over-wrapper').simulate('click');

      expect(toggleOverlayMock).toHaveBeenCalled();
    });
  });
});
