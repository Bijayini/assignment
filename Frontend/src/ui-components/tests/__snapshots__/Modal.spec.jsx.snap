// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match with snapshot 1`] = `
"<Portal containerInfo={{...}}>
  <div className=\\"pop-over-wrapper\\" onClick={[Function]} role=\\"alertdialog\\">
    <div className=\\"pop-over\\">
      <a href=\\"#\\" className=\\"close-icon\\" onClick={[Function: mockConstructor]}>
        Ã—
      </a>
      <div className=\\"some-name\\">
        Some text
      </div>
      <div className=\\"button-wrapper\\">
        <Button type=\\"button\\" onClick={[Function: mockConstructor]} variant=\\"primary\\" size=\\"medium\\" classList=\\"left-button\\">
          Some Button
        </Button>
        <Button type=\\"button\\" onClick={[Function: mockConstructor]} variant=\\"secondary\\" size=\\"medium\\" classList=\\"\\">
          cancel
        </Button>
      </div>
    </div>
  </div>
</Portal>"
`;
